fixed crash with search extension

simonhong opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
fix: https://github.com/brave/brave-browser/issues/15224
fix: https://github.com/brave/brave-browser/issues/10601
fix: https://github.com/brave/brave-browser/issues/14218

With this PR, brave uses search provider from extension by default on all platform like chromium does.

Crash comes from absense of exension search provider data in private profile's TemplateURLService because
we are using different TemplateURLService for private profile. When search provider extension is loaded/installed, browser doesn't notify about this adding to private profile's TUS.
To fix this, each SearchEngineProviderService sets extension's template url data to its service when normal profile's search provider comes from extension.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
`npm run test brave_browser_tests -- --filter=*ExtensionSearchProviderWithPrivateWindow*`

1. Launch browser
2. Install ecosia extension - https://chrome.google.com/webstore/detail/ecosia-the-search-engine/eedlgdlajadkbbjoobobefphmfkcchfk/related
3. Check Normal window's default search provider is ecosia
4. Open private window and check ecosia is default search provider
5. Turn ddg button on doesn't change search provider - still using ecosia
6. Turn ddg button off doesn't change search provider - still using ecosia
7. Goto ecosia extension details and toggle `Allow in private` repeatedly
8. Check changing `Allow in privatge` option doesn't affect private window's search provider - still using ecosia
9. Disable ecosia extension and check normal and private window uses default search provider
10. Toggle DDG button on in private NTP and check search provider is changed to DDG

simonhong commented:
Hmm, it's really hard to control search provider by setting `kDefaultSearchProviderDataPrefName` manually whenever search provider extension's setting is changed.

simonhong commented on browser/search_engines/tor_window_search_engine_provider_service.cc:
```diff
@@ -5,8 +5,11 @@
 
 #include "brave/browser/search_engines/tor_window_search_engine_provider_service.h"
 
+#include "brave/common/pref_names.h"
```
delete this.

simonhong commented:
Kindly ping :)

iefremov commented:
I'll defer to @goodov

goodov commented on browser/search_engines/search_engine_provider_service.cc:
```diff
@@ -76,3 +85,47 @@ void SearchEngineProviderService::ChangeToNormalWindowSearchEngineProvider() {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  DCHECK(ShouldUseExtensionSearchProvider());
+
+  const auto* extension_provider_url =
```
`DCHECK(extension_provider_url);`

simonhong replied:
done.

goodov commented on browser/search_engines/search_engine_provider_service.h:
```diff
@@ -30,21 +35,22 @@ class SearchEngineProviderService : public KeyedService {
   void ChangeToAlternativeSearchEngineProvider();
   void ChangeToNormalWindowSearchEngineProvider();
 
+  bool ShouldUseExtensionSearchProvider();
```
`const`?

simonhong replied:
done.

goodov commented on browser/search_engines/tor_window_search_engine_provider_service.cc:
```diff
@@ -15,16 +17,34 @@ TorWindowSearchEngineProviderService(Profile* otr_profile)
-  TemplateURL provider_url(*provider_data);
-  otr_template_url_service_->SetUserSelectedDefaultSearchProvider(
-      &provider_url);
+  alternative_search_engine_url_for_tor_.reset(new TemplateURL(*provider_data));
```
`make_unique`

simonhong replied:
done.

goodov commented on browser/search_engines/tor_window_search_engine_provider_service.cc:
```diff
@@ -15,16 +17,34 @@ TorWindowSearchEngineProviderService(Profile* otr_profile)
+
+  ConfigureSearchEngineProvider();
+
+  observation_.Observe(original_template_url_service_);
```
also should reset on `Shutdown`.

simonhong replied:
done.

goodov commented on browser/search_engines/tor_window_search_engine_provider_service.h:
```diff
@@ -8,23 +8,34 @@
+  base::ScopedObservation<TemplateURLService, TemplateURLServiceObserver>
+      observation_{this};
 
   DISALLOW_COPY_AND_ASSIGN(TorWindowSearchEngineProviderService);
```
we can also replace that with `= delete` and remove `#include macros.h`.

simonhong replied:
done.

goodov commented on components/search_engines/pref_names.h:
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
what about extending `search_engines_pref_names.h`? less conflict.

simonhong replied:
done.

goodov replied:
I thought about extending chromium file via `chromium_src/components/search_engines/search_engines_pref_names.h`.
That way you don't need to add `brave_components_search_engines_sources` in `.gn`.

simonhong replied:
@goodov  Oops! :) Addressed by https://github.com/brave/brave-core/pull/8626/commits/131c9aabd5e5bef4c60264b53107baac839dcab8

goodov commented on patches/components-search_engines-default_search_manager.cc.patch:
```diff
@@ -0,0 +1,12 @@
+   // Check if default search preference is overriden by extension.
+   if (pref->IsExtensionControlled()) {
+     extension_default_search_ = std::move(turl_data);
++  LOADDEFAULTSEARCHENGINEFROMPREFS_BRAVE
```
indent doesn't seem right.

goodov commented on browser/search_engines/search_engine_provider_service.cc:
```diff
@@ -76,3 +85,47 @@ void SearchEngineProviderService::ChangeToNormalWindowSearchEngineProvider() {
+    const TemplateURL* url) {
+  DCHECK_NE(TemplateURL::NORMAL, url->type());
+  for (const auto* turl : otr_template_url_service_->GetTemplateURLs()) {
+    if (url->type() == turl->type() &&
```
`DCHECK(url)` and `DCHECK(turl)` would be nice.

simonhong replied:
done.

simonhong commented on browser/search_engines/private_window_search_engine_provider_service.cc:
```diff
@@ -5,47 +5,63 @@
   // used.
-  original_template_url_service_->AddObserver(this);
-  ConfigureSearchEngineProvider();
+  observation_.Observe(original_template_url_service_);
```
done.

simonhong commented on browser/search_engines/search_engine_provider_service.h:
```diff
@@ -30,21 +35,22 @@ class SearchEngineProviderService : public KeyedService {
+  std::unique_ptr<TemplateURL> alternative_search_engine_url_;
   BooleanPrefMember use_alternative_search_engine_provider_;
-
-  DISALLOW_COPY_AND_ASSIGN(SearchEngineProviderService);
```
done.

simonhong commented on browser/search_engines/search_engine_provider_service.cc:
```diff
@@ -49,8 +59,7 @@ SearchEngineProviderService::SearchEngineProviderService(
 
-SearchEngineProviderService::~SearchEngineProviderService() {
-}
+SearchEngineProviderService::~SearchEngineProviderService() = default;
```
Hmm, I think we don't need to use it here.
This class uses one var (`BooleanPrefMember`) for monitoring prefs change, but I think it's fine to leave it to destroy in dtor.

simonhong reviewed (commented):
@goodov Thanks for review. All addressed!

goodov commented on components/search_engines/search_engines_pref_names.h:
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
I thought about extending chromium file via `chromium_src/components/search_engines/search_engines_pref_names.h`.
That way you don't need to add `brave_components_search_engines_sources` in `.gn`.

goodov commented:
got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.

simonhong commented:
@goodov Thanks for sharing DCHECK failure. Failure from `OnFaviconUpdated` is very new to me.
Will investigate this more. Thanks again! 👍

simonhong commented:
> @simonhong
> got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.

@goodov  Fixed by https://github.com/brave/brave-core/pull/8626/commits/d091ba76b8d588a45ce34301e1cd715b2a07d809.  PTAL again!

goodov commented:
> > @simonhong
> > got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> > reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.
> 
> @goodov Fixed by [d091ba7](https://github.com/brave/brave-core/commit/d091ba76b8d588a45ce34301e1cd715b2a07d809). PTAL again!

Still doesn't seem right. After disabling the Extension search engine via Settings and enabling extension manually I got this error: https://gist.github.com/goodov/58a34a9ed19350984a9454b6522cae4d

simonhong commented:
> > > @simonhong
> > > got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> > > reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.
> > 
> > 
> > @goodov Fixed by [d091ba7](https://github.com/brave/brave-core/commit/d091ba76b8d588a45ce34301e1cd715b2a07d809). PTAL again!
> 
> Still doesn't seem right. After disabling the Extension search engine via Settings and enabling extension manually I got this error: https://gist.github.com/goodov/58a34a9ed19350984a9454b6522cae4d

Fixed by https://github.com/brave/brave-core/pull/8626/files#diff-303c4f4e4e368bfb2681efc54ff35e1412d362740939fc50c7f93b9b92bcca2cR139. Changed to setting non-empty sync_guid. This PR will work :) Thanks for catching this.

And I'm trying another approach - This PR added new patches because we didn't make `kDefaultSearchProviderDataPrefName` as a extension controlled prefs. but I'm trying to set data to `kDefaultSearchProviderDataPrefName` as a extension controlled pref. If I do that, we can fix this issue w/o patching.

goodov commented:
latest version seems fine, no any visible issues and the logic works great!

> And I'm trying another approach - This PR added new patches because we didn't make `kDefaultSearchProviderDataPrefName` as a extension controlled prefs. but I'm trying to set data to `kDefaultSearchProviderDataPrefName` as a extension controlled pref. If I can do that, we can fix this issue w/o patching.

so should we wait for solution without patching? :)

bridiver reviewed (approved):
approved, but please follow-up as discussed to see if we can get rid of the patch

simonhong commented:
Merged because only unrelated brave ads/rewards browser tests were failed.
